File: replay-missed-stripe-webhook-events/python/replay_events.py
"""Replay Stripe webhook events WooCommerce missed during downtime.
Idempotent. Run once after an outage, or on a schedule as a safety net.

Guide: https://www.allanninal.dev/woocommerce/replay-missed-stripe-webhook-events/
"""
import os
import time
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("replay_events")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_HOURS = int(os.environ.get("LOOKBACK_HOURS", "120"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

PAID_STATUSES = {"processing", "completed"}


def extract_action(event):
    obj = event.get("data", {}).get("object", {})
    order_id = (obj.get("metadata") or {}).get("order_id")
    if not order_id:
        return None
    if event["type"] == "payment_intent.succeeded":
        return ("complete", order_id, obj)
    if event["type"] == "charge.refunded":
        return ("refund", order_id, obj)
    return None


def undelivered_events(lookback_hours):
    since = int(time.time()) - lookback_hours * 3600
    return stripe.Event.list(
        limit=100, created={"gte": since}, delivery_success=False,
        types=["payment_intent.succeeded", "charge.refunded"],
    ).auto_paging_iter()


def get_order(order_id):
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def put_order(order_id, body):
    requests.put(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", json=body, auth=AUTH, timeout=30).raise_for_status()


def add_note(order_id, note):
    requests.post(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/notes",
                  json={"note": note}, auth=AUTH, timeout=30).raise_for_status()


def sync_refund(order, charge):
    order_total_minor = round(float(order["total"]) * 100)
    stripe_refunded = charge.get("amount_refunded", 0)
    wc_refunded = sum(round(abs(float(r["total"])) * 100) for r in order.get("refunds", []))
    missing = stripe_refunded - wc_refunded
    if missing > 1:
        requests.post(f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}/refunds",
                      json={"amount": f"{missing / 100:.2f}", "api_refund": False,
                            "reason": "Replayed a missed Stripe refund event."},
                      auth=AUTH, timeout=30).raise_for_status()
        if stripe_refunded >= order_total_minor and order["status"] != "refunded":
            put_order(order["id"], {"status": "refunded"})


def apply_event(action, order_id, obj):
    order = get_order(order_id)
    if not order:
        return False
    if action == "complete" and order["status"] not in PAID_STATUSES:
        charge_id = obj.get("latest_charge") or obj["id"]
        put_order(order_id, {"status": "processing", "transaction_id": charge_id})
        add_note(order_id, "Replayed a missed Stripe payment event. Marked processing.")
        return True
    if action == "refund":
        sync_refund(order, obj)
        return True
    return False


def run():
    seen = set()
    applied = 0
    for event in undelivered_events(LOOKBACK_HOURS):
        if event["id"] in seen:
            continue
        seen.add(event["id"])
        parsed = extract_action(event)
        if not parsed:
            continue
        action, order_id, obj = parsed
        log.info("Event %s -> %s order %s. %s", event["id"], action, order_id, "dry run" if DRY_RUN else "applying")
        if not DRY_RUN and apply_event(action, order_id, obj):
            applied += 1
    log.info("Done. %d event(s) %s.", applied, "found" if DRY_RUN else "reapplied")


if __name__ == "__main__":
    run()

File: replay-missed-stripe-webhook-events/node/replay-events.js
/**
 * Replay Stripe webhook events WooCommerce missed during downtime.
 * Idempotent. Run once after an outage, or on a schedule as a safety net.
 *
 * Guide: https://www.allanninal.dev/woocommerce/replay-missed-stripe-webhook-events/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_HOURS = Number(process.env.LOOKBACK_HOURS || 120);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const PAID_STATUSES = new Set(["processing", "completed"]);

export function extractAction(event) {
  const obj = (event.data && event.data.object) || {};
  const orderId = (obj.metadata || {}).order_id;
  if (!orderId) return null;
  if (event.type === "payment_intent.succeeded") return { action: "complete", orderId, obj };
  if (event.type === "charge.refunded") return { action: "refund", orderId, obj };
  return null;
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function* undeliveredEvents(lookbackHours) {
  const since = Math.floor(Date.now() / 1000) - lookbackHours * 3600;
  const params = { limit: 100, created: { gte: since }, delivery_success: false,
                   types: ["payment_intent.succeeded", "charge.refunded"] };
  for await (const event of stripe.events.list(params)) yield event;
}

async function syncRefund(order, charge) {
  const orderTotalMinor = Math.round(parseFloat(order.total) * 100);
  const stripeRefunded = charge.amount_refunded || 0;
  const wcRefunded = (order.refunds || []).reduce((s, r) => s + Math.round(Math.abs(parseFloat(r.total)) * 100), 0);
  const missing = stripeRefunded - wcRefunded;
  if (missing > 1) {
    await woo(`/orders/${order.id}/refunds`, { method: "POST", body: JSON.stringify({ amount: (missing / 100).toFixed(2), api_refund: false, reason: "Replayed a missed Stripe refund event." }) });
    if (stripeRefunded >= orderTotalMinor && order.status !== "refunded") {
      await woo(`/orders/${order.id}`, { method: "PUT", body: JSON.stringify({ status: "refunded" }) });
    }
  }
}

async function applyEvent(action, orderId, obj) {
  const order = await woo(`/orders/${orderId}`);
  if (!order) return false;
  if (action === "complete" && !PAID_STATUSES.has(order.status)) {
    const chargeId = obj.latest_charge || obj.id;
    await woo(`/orders/${orderId}`, { method: "PUT", body: JSON.stringify({ status: "processing", transaction_id: chargeId }) });
    await woo(`/orders/${orderId}/notes`, { method: "POST", body: JSON.stringify({ note: "Replayed a missed Stripe payment event. Marked processing." }) });
    return true;
  }
  if (action === "refund") { await syncRefund(order, obj); return true; }
  return false;
}

export async function run() {
  const seen = new Set();
  let applied = 0;
  for await (const event of undeliveredEvents(LOOKBACK_HOURS)) {
    if (seen.has(event.id)) continue;
    seen.add(event.id);
    const parsed = extractAction(event);
    if (!parsed) continue;
    const { action, orderId, obj } = parsed;
    console.log(`Event ${event.id} -> ${action} order ${orderId}. ${DRY_RUN ? "dry run" : "applying"}`);
    if (!DRY_RUN && await applyEvent(action, orderId, obj)) applied++;
  }
  console.log(`Done. ${applied} event(s) ${DRY_RUN ? "found" : "reapplied"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
